auth_core/pkce.py

```diff
@@ -0,0 +1,50 @@
+import base64
+import hashlib
+import httpx
+import logging
+import secrets
+
+from urllib.parse import urljoin
+
+from .token_data import TokenData
+
+
+logger = logging.getLogger("gel_auth_core")
+
+
+class PKCE:
+    def __init__(self, verifier: str, *, base_url: str):
+        self.base_url = base_url
+        self.verifier = verifier
+        self.challenge = (
+            base64.urlsafe_b64encode(hashlib.sha256(verifier.encode()).digest())
+            .rstrip(b"=")
+            .decode()
+        )
+
+    async def exchange_code_for_token(self, code: str) -> TokenData:
+        async with httpx.AsyncClient() as http_client:
+            url = urljoin(self.base_url, "token")
+            logger.info(f"Exchanging code for token: {url}")
+            token_response = await http_client.get(
+                url,
+                params={
+                    "code": code,
+                    "verifier": self.verifier,
+                },
+            )
+
+            logger.info(f"Token response: {token_response.text}")
```

Comment:
```suggestion
```

or

```suggestion
            logger.debug(f"Token response: {token_response.text}")
```

As we don't have a generic way to redact sensitive information from logging, it's probably better not to show the raw tokens by default.